### ISSUE TITLE
[data request] Deepmind's PG-19

### ISSUE THREAD

**Issue by afrozenator** [tensorflow/datasets#2048]
* Name of dataset: PG-19
* URL of dataset: https://github.com/deepmind/pg19
* License of dataset: Apache License 2.0
* Short description of dataset and use case(s): This repository contains the PG-19 language modeling benchmark. It includes a set of books extracted from the Project Gutenberg books library, that were published before 1919. It also contains metadata of book titles and publication dates. PG-19 is over double the size of the Billion Word benchmark and contains documents that are 20X longer, on average, than the WikiText long-range language modelling benchmark.

https://github.com/deepmind/pg19

Folks who would also like to see this dataset in `tensorflow/datasets`, please thumbs-up so the developers can know which requests to prioritize.

And if you'd like to contribute the dataset (thank you!), see our [guide to adding a dataset](https://github.com/tensorflow/datasets/blob/master/docs/add_dataset.md).

**Comment by afrozenator** [comment]
@lukaszkaiser - please upvote?

**Comment by lukaszkaiser** [comment]
It's a very nice dataset, would be great to have it on TFDS for more impact!

**Comment by acharles7** [comment]
@afrozenator @lukaszkaiser Can I work on this?

**Comment by lukaszkaiser** [comment]
Sure!! Let us know if we can help in any way!

**Comment by acharles7** [comment]
Thanks @lukaszkaiser 
Is it fine if I put this data on the text section of the dataset?
Do we only need metadata for this dataset? because if not then this dataset is of > 10GB.
Please let me know what you really want in this dataset
Thanks.

**Comment by lukaszkaiser** [comment]
I think we want the whole dataset - with all the data and all splits, right? So anyone using TFDS can run experiments with this dataset same as with any other language modelling one. I guess so at least - I think that'd be the most beneficial for TFDS and all users, or am I missing something?

**Comment by acharles7** [comment]
Thank you, I'm working on it. I'll let you know if I have any queries. I will make PR soon.

**Comment by lukaszkaiser** [comment]
Great news, looking forward to it!! :)

**Comment by afrozenator** [comment]
Hi @acharles7 -- In TFDS you write to code to process the dataset and store as tf-records, while you are iterating, maybe you can just process one or two files from the train split and one or two from validation split etc just to make things fast on your end.

In the eventual PR you can remove these constraints and then anyone who needs it can build the dataset on their own.

Let me know if you have questions!

And thanks for taking it up!

**Comment by afrozenator** [comment]
I'll assign this issue to you at the moment, @acharles7 - many thanks again!

**Comment by acharles7** [comment]
Thank you @afrozenator I'll let you know if I have any doubts.

**Comment by acharles7** [comment]
@lukaszkaiser @afrozenator I got this error when running `download_prepare.py` script

> `tensorflow.python.framework.errors_impl.ResourceExhaustedError:` and sometimes HTTPconnection lost 

As this dataset is of size >10 GB, and I'm saving the dataset locally. it might be the case of this error. Should I look into the Apache beam? As documentation suggests beam for TB sized dataset.

here's the code for the `my_dataset.py`
```
DOWNLOAD_URLS = {
    'metadata': _METADATA_URL,
    'train': [list of urls],
    'test': [list of urls],
    'val': [list of urls]
}


class Pg19(tfds.core.GeneratorBasedBuilder):
  """This dataset contains the PG-19 language modeling benchmark"""

  VERSION = tfds.core.Version('0.1.0')

  def _info(self):

    return tfds.core.DatasetInfo(
        builder=self,
        # This is the description that will appear on the datasets page.
        description=_DESCRIPTION,
        # tfds.features.FeatureConnectors
        features=tfds.features.FeaturesDict({
            'book_text': tfds.features.Text(),
            # 'metadata': {
            #     'book_id': tf.int32,
            #     'book_title': tf.string,
            #     'publication_date': tf.string,
            #     'book_link': tf.string
            # },
        }),
        supervised_keys=None,
        homepage='https://github.com/deepmind/pg19',
        citation=_CITATION,
    )

  def _split_generators(self, dl_manager):
    """Returns SplitGenerators."""
    dl_dir = dl_manager.download(_DOWNLOAD_URLS)
    meta_data = tf.io.gfile.GFile(dl_dir["metadata"]).read().splitlines()

    return [
        tfds.core.SplitGenerator(
            name=tfds.Split.TRAIN,
            gen_kwargs={
                "metadata": meta_data,
                "filepaths": dl_dir['train']},
        ),
        tfds.core.SplitGenerator(
            name=tfds.Split.VALIDATION,
            gen_kwargs={
                "metadata": meta_data,
                "filepaths": dl_dir['val']},
        ),
        tfds.core.SplitGenerator(
            name=tfds.Split.TEST,
            gen_kwargs={
                "metadata": meta_data,
                "filepaths": dl_dir['test']},
        ),
    ]

  def _generate_examples(self, filepaths, metadata):
    """Yields examples."""

    metadata_dict = collections.defaultdict(list)

    for row in metadata:
      features = row.split(',')
      metadata_dict['book_id'].append(int(features[0]))
      metadata_dict['book_title'].append(str(features[1]))
      metadata_dict['publication_date'].append(str(features[2]))
      metadata_dict['book_link'].append(str(features[3]))

    for index, file in enumerate(filepaths):
      with tf.io.gfile.GFile(file, 'r') as f:
        text = f.read().strip()
        yield index, {"book_text": text}
```
Please suggest any better approach for this or Let me know if I'm doing wrong somewhere.
TIA

**Comment by afrozenator** [comment]
Hi @acharles7 - I think this looks good, have you verified that this runs on a small sample of the dataset?

i.e. in `_generate_examples` instead of going over all files here:
```
for index, file in enumerate(filepaths):
...
```

What if you break after one file, i.e.
```
for index, file in enumerate(filepaths):
  # do stuff and yield example
  ...
  # DO NOT SUBMIT - process only one file to test.
 break
```

Is it neat if we can get the metadata too?

I don't think we'll need Beam for this dataset, 10Gb should be easy.

As again, please just iterate on one file per split to nail down other aspects.

Does that help?

Thanks!

**Comment by acharles7** [comment]
Thank you @afrozenator 
Yes, I have generated `dataset_info.json` file for validation and test set only.
Can you help me with the `metadata.csv` file, How to yield metadata and textfile content together?
I'll try other steps mentioned by you and get back to you.

**Comment by acharles7** [comment]
I generated this output after iterating only 1 one file for each split.
which means it works properly. right?

```
location {                                                                                                                      
  urls: "https://github.com/deepmind/pg19"                                                                                      
}                                                                                                                               
splits {                                                                                                                        
  name: "test"                                                                                                                  
  shard_lengths: 1                                                                                                              
  num_bytes: 253903                                                                                                             
}                                                                                                                               
splits {                                                                                                                        
  name: "train"                                                                                                                 
  shard_lengths: 1                                                                                                              
  num_bytes: 4432491                                                                                                            
}                                                                                                                               
splits {                                                                                                                        
  name: "validation"                                                                                                            
  shard_lengths: 1                                                                                                              
  num_bytes: 69018                                                                                                              
}                                                                                                                               
version: "0.1.0"                                                                                                                
download_size: 11739944217
```

**Comment by afrozenator** [comment]
So far so good.

For metadata:

I see that in `_generate_examples` you create a dictionary (`metadata_dict`) out of the `meta_data` list (and you create `meta_data` list in `_split_generators`) correct?

Instead let us move the dictionary generation to `_split_generators` entirely, so that each `_generate_examples` call doesn't need to do a lot of processing.

Also let's change the dictionary format to be the following (this will allow us to find the metadata easily in `_generate_examples`:
dict metadata_dict :   book_id -> (book_title, publication_date, book_link)

This will just be a slight change to your dict generation code.

Then in `_generate_examples` we can simply do:

```
    for index, file in enumerate(filepaths):
     # Ex: file is something like: "10146.txt"
     book_id = file.rstrip(".txt")
      with tf.io.gfile.GFile(file, 'r') as f:
        text = f.read().strip()
        metadata = metadata_dict[book_id]
        yield index, {"text": text, "book_id": book_id, "book_title": metadata[0], "publication_date": metadata[1], "book_link": metadata[2]}
```

Makes sense?

**Comment by acharles7** [comment]
Yes, gotcha. 
Thank you @afrozenator

**Comment by acharles7** [comment]
I have one query related to `book_id`, the book_id related information is lost after it downloaded using `dl_manager.download(urls)`, This is what it looks like 
```
for index, file in enumerate(filepaths):
     # Ex: file is something like: "deepmind-gutenberg_test_10356r0tX01hEhYitvG2pYsrWVb9YaudG26IfYulW022XyI4.txt"
     # We can't strip and retrieve book_id from above file format
     book_id = file.rstrip(".txt") 
      with tf.io.gfile.GFile(file, 'r') as f:
        text = f.read().strip()
        metadata = metadata_dict[book_id]
        yield index, {"text": text, "book_id": book_id, "book_title": metadata[0], "publication_date": metadata[1], "book_link": metadata[2]}
```
>"deepmind-gutenberg_test_**10356**r0tX01hEhYitvG2pYsrWVb9YaudG26IfYulW022XyI4.txt"
_The bold text is book_id_

I can't find a way to retrieve `book_id` from the downloaded file format that we got in `_generate_examples`. 
Do you have any idea to retrieve `book_id` in `_split_generators` or `_generate_examples` from downloaded files, So we can easily map `book_id` with its metadata from `metadata_dict`?

**Comment by afrozenator** [comment]
Looking into it.

**Comment by Conchylicultor** [comment]
The data seems to be hosted on GCS. It means that you could read directly the original files:

```
for filename in tf.io.gfile.listdir('gs://deepmind-gutenberg/test'):
  ...
```

This way, you don't have to download the file and deal with name mangling.

**Comment by acharles7** [comment]
@Conchylicultor thank you.
I'll try this too.

**Comment by acharles7** [comment]
@afrozenator I have completed the coding. here's the snippet of it. Let me know if it requires changes. I have run `download_and_prepare ` for only on test and validation set and generated `dataset_info.json`. Is it fine if I put `dataset_info.json` with partial information (without train set split) at  the time of PR?


```
_DATA_DIR = 'gs://deepmind-gutenberg'

def _split_generators(self, dl_manager):
    """Returns SplitGenerators."""
    del dl_manager

    metadata_dict = dict()
    metadata_path = os.path.join(_DATA_DIR, "metadata.csv")
    metadata = tf.io.gfile.GFile(metadata_path).read().splitlines()

    for row in metadata:
      row_split = row.split(',')
      # book_id: [book_title, publication_date, book_link]
      metadata_dict[int(row_split[0])] = row_split[1:]

    return [
        tfds.core.SplitGenerator(
            name=tfds.Split.TRAIN,
            gen_kwargs={
                "metadata": metadata_dict,
                "filepath": os.path.join(_DATA_DIR, "train")},
        ),
        ....
    ]
```

```
def _generate_examples(self, filepath, metadata):
    """Yields examples."""
    for index, file in enumerate(tf.io.gfile.listdir(filepath)):
      book_id = int(file.rstrip('.txt'))
      book_data = metadata[book_id]
      path = os.path.join(filepath, file)
      with tf.io.gfile.GFile(path, 'r') as f:
        text = f.read().strip()
        yield index, {
            "book_text": text,
            "book_id": book_id,
            "book_title": book_data[0],
            "publication_date": book_data[1],
            "book_link": book_data[2],
            }
```

**Comment by afrozenator** [comment]
Thanks @acharles7 please make the PR!

**Comment by acharles7** [comment]
@afrozenator @lukaszkaiser @Conchylicultor Thank you for your help in getting this PR done. You guys always inspire me to do more _open source contributions_

**Comment by afrozenator** [comment]
Many many thanks for doing this @acharles7 !
